### PR TITLE
Handle duplicate contacts, closes #87

### LINKED ISSUE #87: Duplicate contacts resolvement
As a user I want to choose if I want to add a newly received public key to a new or to an existing contact so that I don't get duplicate contacts after receiving 5 public keys from the same person.

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: #87 
- Related Pull Requests: -

* * *

## What
This Pull Request adds to the repository a correct handling of newly received contacts with the name of a contact that the user already has.

- SharedActivityMethods.java handles all shared methods by the activities, i.e. duplicate contacts handlers
- If the contact is duplicate, the user can choose between adding the new public keys to the existing contact or to create a new contact

## Why
This Pull Request is needed because the behavior for duplicate contacts is now undefined

## How
This feature can be viewed/tested within the project by receiving a duplicate contact and verifying that the SweetAlerts work correctly

## Alternative implementation
Other implementations that I've have considered are

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                .showCancelButton(true)
+                .setCancelClickListener(new SharedActivityMethods.CreateNewContactClickListener(activity, database, contact))
+                .setConfirmClickListener(new SharedActivityMethods.AddPublicKeyToContactsClickListener(database, contact))
+                .show();
+    }
+
+    /**
+     * Checks if a name of a given contact exists in the database.
+     *
+     * @param contact A contact object
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: {link @Contact}

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+import cn.pedant.SweetAlert.SweetAlertDialog;
+
+/**
+ * Contains shared functionality between classes to reduce code duplication
+ */
+// Suppressed because we have an empty enum
+@SuppressWarnings("checkstyle:nowhitespacebefore")
+enum SharedActivityMethods {
+    ;
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: Should this ";" be there?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        AddPublicKeyToContactsClickListener(final IDatabase database, final Contact contact) {
+            this.database = database;
+            this.contact = contact;
+        }
+
+        @Override
+        public void onClick(final SweetAlertDialog sweetAlertDialog) {
+            try {
+                final Contact existingContact = this.database.getContactWithName(this.contact.getName());
+                existingContact.addKeys(this.contact.getKeys());
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: But what happens when the existing contact also already has the public key? It looks like nothing happens

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    private static void handleExistingContact(final IDatabase database, final Activity activity, final Contact contact) {
+        new SweetAlertDialog(activity, SweetAlertDialog.WARNING_TYPE)
+                .setTitleText(activity.getString(R.string.contact_already_exists))
+                .setContentText(String.format(activity.getString(R.string.contact_exists), contact.getName()))
+                .setConfirmText(activity.getString(R.string.add_public_key_to_contact))
+                .setCancelText(activity.getString(R.string.create_new_contact))
+                .showCancelButton(true)
+                .setCancelClickListener(new SharedActivityMethods.CreateNewContactClickListener(activity, database, contact))
+                .setConfirmClickListener(new SharedActivityMethods.AddPublicKeyToContactsClickListener(database, contact))
+                .show();
```
Comment (clenrock) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: Should we also give a popup when it was added successfully after it?

File: app/src/main/res/values-nl/strings.xml [hunk 5]
```diff
@@ -95,10 +100,9 @@
     <string name="scan_qr_code">Scan QR code</string>
     <string name="show_qr_code">Toon QR code</string>
     <string name="contact_set_name">Geef contact een naam</string>
-    <string name="contact_exists_message">Contact bestaat al, geef aub een unieke naam</string>
+    <string name="contact_exists_give_unique_name">Contact bestaat al, geef aub een unieke naam</string>
```
Comment (clenrock) [inline app/src/main/res/values-nl/strings.xml]: Is al uit een oude PR zie ik, maar kun je 'aub' veranderen in "alstublieft" of "a.u.b."

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java (hunk 2 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: Yes, because it's an enum just to prevent instantiation and does not actually contain any elements.

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java (hunk 4 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: I don't think so, because if I want to save a Word file with the same name as an existing file, then it asks me if I want to rename it. If I do that I won't get a notification afterwards that successfully a new file with the name I entered is created

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java (hunk 3 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: Fixed that

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 6]
```diff
@@ -41,6 +42,7 @@
     private Button doneButton;
     private ArrayList<SingleTap> tapCombination;
     private IServiceLocator serviceLocator;
+    private IDatabase database;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Not sure if you want this as an attribute since you're using it only once 🤔

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 7]
```diff
@@ -100,7 +102,7 @@ protected void onActivityResult(final int requestCode, final int resultCode, fin
     protected void onStart() {
         super.onStart();
         if (this.dataReceived != null && this.tapCombination != null) {
-            evaluateData();
+            this.goToMainActivity();
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Do you no longer want to evaluate the data?

File: app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.Set;
+
+import cn.pedant.SweetAlert.SweetAlertDialog;
+
+/**
+ * Contains shared functionality between classes to reduce code duplication
+ */
+// Suppressed because we have an empty enum
+@SuppressWarnings("checkstyle:nowhitespacebefore")
+enum SharedActivityMethods {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/SharedActivityMethods.java]: I think this enum should only be used for on contact receive, so maybe change its name to reflect that?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java [hunk 9]
```diff
@@ -38,29 +40,56 @@ public Contact(final String name, final IKey key) {
...
+        final List<IKey> otherKeys = otherContact.getKeys();
+        if (otherKeys.size() != this.keys.size()) {
+            return false;
+        }
+        for (int i = 0; i < this.keys.size(); i++) {
+            if (!this.keys.get(i).equals(otherKeys.get(i))) {
+                return false;
+            }
+        }
+        return this.name.equals(otherContact.getName());
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Why did you change all this ?

File: app/src/main/res/values-nl/strings.xml [hunk 10]
```diff
@@ -33,7 +33,7 @@
     <string name="get_started">Laten we beginnen!</string>
     <string name="go_back">Ga terug</string>
     <string name="invalid_name">Ongeldige naam</string>
-    <string name="key_type" formatted="false">Type sleutel: %s</string>
+    <string name="key_type">Type sleutel: %1$s</string>
```
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: Why did you remove `formatted="false"`?

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Because you were comparing a list with a list. If the other contact has a list with exactly the same size and contacts but with a different memory address, the old equals method would return false

File: app/src/main/res/values-nl/strings.xml (hunk 10 above)
Comment (jverbraeken) [inline app/src/main/res/values-nl/strings.xml]: Because it is formatted xD

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 7 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: I renamed evaluateData to goToMainActivity

File: app/src/main/res/values-nl/strings.xml (hunk 10 above)
Comment (ericcornelissen) [inline app/src/main/res/values-nl/strings.xml]: Okey, misunderstood what the formatted attribute does 😅

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Not if the `Contacts` equals works correctly, right?

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 7 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Okay, then `evaluateData` maybe was a bad name... Can you edit the log in `goToMainActivity` (since it doesn't seem to be evaluating any data)?

Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Done

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: This is the Contact equals, and it didn't worked correctly as I explained above

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 7 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Thx 😄

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Ow srry, was a little confused about in what file this `equals` was 😅 

I'm still not sure what you mean though 😕 The equals method of strings (i.e. for the name) doesn't depend on the memory location and the some goes for the list of `IKey`s?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 6 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Missed those 😅

File: app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java (hunk 9 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Brainfried, you're right. I don't know what I was thinking

Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java]: Haha, np 😄

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java [hunk 11]
```diff
@@ -149,10 +140,19 @@ public void onNewDataReceivedEvent(final NewDataReceivedEvent event) {
...
      * Can be called by a button to cancel the pairing
+     *
      * @param view The view that called this method
      */
     public void cancelWaiting(final View view) {
         this.setResult(ConstantKeywords.CANCEL_PAIRING_RESULT_CODE);
         this.finish();
     }
+
+    private void goToMainActivity() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Wasn't there before, but could you add JavaDoc 😇

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 12]
```diff
@@ -172,6 +174,15 @@ public void onStopRecordingClick(final View v) {
...
         this.doneButton.setVisibility(View.VISIBLE);
+        if (this.taps.size() < MINIMUM_TAPS) {
+            this.taps.clear();
+            this.doneButton.setVisibility(View.GONE);
+            new SweetAlertDialog(this, SweetAlertDialog.ERROR_TYPE)
+                    .setTitleText(this.getString(R.string.too_few_taps_title))
+                    .setContentText(this.getString(R.string.too_few_taps_description))
+                    .setConfirmText(this.getString(R.string.dialog_ok))
+                    .show();
+        }
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Can you put `this.doneButton.setVisibility(View.VISIBLE);` in an `else` branch?

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java [hunk 13]
```diff
@@ -191,4 +202,145 @@ public void onNewDataReceivedEvent(final NewDataReceivedEvent event) {
...
+                clusterCenterMean1 = newClusterCenters.getLeft();
+                clusterCenterMean2 = newClusterCenters.getRight();
+            }
+
+            final long breakpoint = this.getBreakpoint();
+            this.intervals = getIntervals(taps);
+            return this.generateKey(breakpoint);
+        }
+
+        private long makeAndReturnFirstClusterMean() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Can you add concise JavaDoc for all these private methods so it is more clear for us what they do? 😄

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 11 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: It's not needed, because this method won't show up in the documentation (because the method is private), so the documentation would be only for the programmer. Because the title is very, very clear I don't think it's needed, but well... Maybe the TA's like comments as well, so I'll do it

File: app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java (hunk 13 above)
Comment (jverbraeken) [inline app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java]: Here I agree with you, because these the purpose of the methods is not very clear indeed

File: app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java (hunk 11 above)
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/activities/WaitActivity.java]: Figured it might be useful as it also sets flags and puts an extra...
